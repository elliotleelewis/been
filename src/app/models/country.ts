export interface Country {
	name: string;
	iso3166: string;
	region: string;
	selected?: boolean;
}
