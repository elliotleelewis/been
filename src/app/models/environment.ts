export interface Environment {
	production: boolean;
	apiKeyMapbox: string;
}
