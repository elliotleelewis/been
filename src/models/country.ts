export interface Country {
	name: string;
	iso3166: string;
	region: string;
	bounds?: [number, number, number, number];
	selected?: boolean;
}
