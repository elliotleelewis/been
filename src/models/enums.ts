export enum MapboxSourceKeys {
	countries = 'countries',
}

export enum MapboxLayerKeys {
	been = 'been',
	buildings = 'buildings',
}
