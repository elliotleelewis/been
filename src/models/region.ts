import type { Country } from './country.ts';

export interface Region {
	name: string;
	values: Country[];
}
