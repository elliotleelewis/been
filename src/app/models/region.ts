import { type Country } from './country';

export interface Region {
	name: string;
	values: Country[];
}
