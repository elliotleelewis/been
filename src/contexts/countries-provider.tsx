import {
	type FC,
	type ReactNode,
	memo,
	useCallback,
	useEffect,
	useMemo,
	useState,
} from 'react';

import { useLocalStorage } from '../hooks/use-local-storage';
import { type Country } from '../models/country';
import { type Region } from '../models/region';
import { regionalizer } from '../utils/regionalizer';

import { CountriesContext } from './countries-context';

const COUNTRIES_STORAGE_KEY = 'APP_COUNTRIES';

interface Props {
	data: readonly Country[];
	children: ReactNode;
}

export const CountriesProvider: FC<Props> = memo(({ data, children }) => {
	const localStorage = useLocalStorage();
	const [selectedCountries, setSelectedCountries] = useState<string[]>(() => {
		const item = localStorage.getItem(COUNTRIES_STORAGE_KEY);
		return item ? (JSON.parse(item) as string[]) : [];
	});
	const [focus, setFocus] = useState<string | null>(null);

	useEffect(() => {
		localStorage.setItem(
			COUNTRIES_STORAGE_KEY,
			JSON.stringify(selectedCountries),
		);
	}, [selectedCountries, localStorage]);

	const countries = useMemo((): readonly Country[] => {
		return data.map((c) => ({
			...c,
			selected: selectedCountries.includes(c.iso3166),
		}));
	}, [data, selectedCountries]);

	const regions = useMemo((): readonly Region[] => {
		return regionalizer(countries);
	}, [countries]);

	const addCountry = useCallback((countryCode: string) => {
		setSelectedCountries((prevCountries) => {
			if (!prevCountries.includes(countryCode)) {
				setFocus(countryCode);
				return [...prevCountries, countryCode];
			}
			return prevCountries;
		});
	}, []);

	const removeCountry = useCallback((countryCode: string) => {
		setSelectedCountries((prevCountries) => {
			setFocus(null);
			const newCountries = prevCountries.filter(
				(code) => code !== countryCode,
			);
			return newCountries;
		});
	}, []);

	const clearCountries = useCallback(() => {
		setSelectedCountries([]);
	}, []);

	const value = useMemo(
		() => ({
			countries,
			regions,
			focus,
			addCountry,
			removeCountry,
			clearCountries,
		}),
		[countries, regions, focus, addCountry, removeCountry, clearCountries],
	);

	return (
		<CountriesContext.Provider value={value}>
			{children}
		</CountriesContext.Provider>
	);
});
CountriesProvider.displayName = 'CountriesProvider';
