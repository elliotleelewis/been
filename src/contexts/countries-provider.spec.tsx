import { render } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { type Country } from '../models/country';

import {
	CountriesContext,
	type CountriesContextType,
} from './countries-context';
import { CountriesProvider } from './countries-provider';

const mockCountries = {
	['CA']: { name: 'Country A', iso3166: 'CA', region: 'Region 1' },
	['CB']: { name: 'Country B', iso3166: 'CB', region: 'Region 2' },
};

describe('CountriesContext', () => {
	beforeEach(() => {
		vi.mock(import('../hooks/use-local-storage'), () => ({
			useLocalStorage: vi.fn(() => [[], vi.fn()] as const),
		}));
	});

	const renderProvider = (data: Record<string, Country> = mockCountries) => {
		let contextValues: CountriesContextType | undefined;

		render(
			<CountriesProvider data={data}>
				<CountriesContext.Consumer>
					{(value) => {
						contextValues = value;
						return null;
					}}
				</CountriesContext.Consumer>
			</CountriesProvider>,
		);

		if (!contextValues) {
			throw new Error('Failed to render provider.');
		}

		return contextValues;
	};

	it('should render provider', () => {
		const result = render(
			<CountriesProvider data={{}}>Hello world!</CountriesProvider>,
		);

		expect(result.asFragment()).toMatchSnapshot();
	});

	it('should render provider with data', () => {
		const result = render(
			<CountriesProvider
				data={{
					['GB']: {
						name: 'United Kingdom',
						iso3166: 'GB',
						region: 'Europe',
					},
				}}
			>
				Hello world!
			</CountriesProvider>,
		);

		expect(result.asFragment()).toMatchSnapshot();
	});

	it('should provide countries with selected property', () => {
		const context = renderProvider();

		expect(context.countries[0]?.selected).toBe(false);
		expect(context.countries[1]?.selected).toBe(false);
	});
});
