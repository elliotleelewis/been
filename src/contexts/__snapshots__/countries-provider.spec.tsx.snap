// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CountriesContext > it should render provider 1`] = `
<DocumentFragment>
  Hello world!
</DocumentFragment>
`;
